### PR TITLE
Add Cookie to dict

### PR DESCRIPTION
**Pull request by Ausli**

### CODE REVIEW

**Comment by yifeikong** [comment]
Actually, you can do this:

```
c = dict(session.cookies)
```

**Comment by Ausli** [comment]
Yes, it's optional, and I thought it would be more convenient to add it, since the requests library has this function as well

**Comment by yifeikong** [comment]
OK, it makes sense to me.

File: curl_cffi/requests/cookies.py [hunk 1]
```diff
@@ -290,6 +290,21 @@ def delete(
         for cookie in remove:
             self.jar.clear(cookie.domain, cookie.path, cookie.name)
 
+    def get_dict(self, domain=None, path=None):
+        """Takes as an argument an optional domain and path and returns a plain
+        old Python dict of name-value pairs of cookies that meet the
+        requirements.
+
+        :rtype: dict
```
Comment (yifeikong) [inline curl_cffi/requests/cookies.py:298]: Change this to typehint.
```suggestion
```

File: curl_cffi/requests/cookies.py [hunk 2]
```diff
@@ -290,6 +290,21 @@ def delete(
         for cookie in remove:
             self.jar.clear(cookie.domain, cookie.path, cookie.name)
 
+    def get_dict(self, domain=None, path=None):
```
Comment (yifeikong) [inline curl_cffi/requests/cookies.py:293]:
```suggestion
    def get_dict(self, domain=None, path=None) -> dict:
```

**Review by yifeikong — CHANGES_REQUESTED** [review]
Also add a simple unittest for this function.

**Comment by Ausli** [comment]
Is there a unit test template for the project, lend me for reference.

**Comment by yifeikong** [comment]
Sorry, there is no such templates, you can take [this](https://github.com/yifeikong/curl_cffi/blob/main/tests/unittest/test_cookies.py) as a reference.

**Comment by Ausli** [comment]
ModuleNotFoundError: No module named 'curl_cffi._wrapper'
How should I get _wrapper.so

**Comment by yifeikong** [comment]
> I get the following error when I run the test_cookies.py file ModuleNotFoundError: No module named 'curl_cffi._wrapper' How should I get _wrapper.so

```
pip install -e .
```

**Comment by Ausli** [comment]
G:\workarea\MyProject\pythonProject\curl_cffi\ffi\shim.h(5): fatal error C1083: �޷��򿪰����ļ�: ��curl/curl.h��: No such file or directory
The compile path is messed up, see what's wrong with it
I suspect the reason is that the program reads the file using GBK encoding, but the file is UTF8 encoded causing the problem

**Comment by Ausli** [comment]
```
(pythonProject) PS G:\workarea\MyProject\pythonProject\curl_cffi> pip install -e .
Obtaining file:///G:/workarea/MyProject/pythonProject/curl_cffi
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> [42 lines of output]
      Traceback (most recent call last):
        File "G:\Env\installer\miniconda\envs\pythonProject\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "G:\Env\installer\miniconda\envs\pythonProject\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "G:\Env\installer\miniconda\envs\pythonProject\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 132, in get_requires_for_build_editable
          return hook(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\build_meta.py", line 441, in get_requires_for_build_editable
          return self.get_requires_for_build_wheel(config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\build_meta.py", line 325, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\build_meta.py", line 295, in _get_build_requires
          self.run_setup()
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\build_meta.py", line 311, in run_setup
          exec(code, locals())
        File "<string>", line 16, in <module>
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\__init__.py", line 103, in setup
          return distutils.core.setup(**attrs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 147, in setup
          _setup_distribution = dist = klass(attrs)
                                       ^^^^^^^^^^^^
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\dist.py", line 303, in __init__
          _Distribution.__init__(self, dist_attrs)
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 283, in __init__
          self.finalize_options()
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\dist.py", line 654, in finalize_options
          ep(self)
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\setuptools\dist.py", line 674, in _finalize_setup_keywords
          ep.load()(self, ep.name, value)
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\cffi\setuptools_ext.py", line 216, in cffi_modules
          add_cffi_module(dist, cffi_module)
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\cffi\setuptools_ext.py", line 49, in add_cffi_module
          execfile(build_file_name, mod_vars)
        File "C:\Users\a5320\AppData\Local\Temp\pip-build-env-gdxz1ugr\overlay\Lib\site-packages\cffi\setuptools_ext.py", line 22, in execfile
          src = f.read()
                ^^^^^^^^
      UnicodeDecodeError: 'gbk' codec can't decode byte 0xa3 in position 1488: illegal multibyte sequence
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

```

**Comment by Ausli** [comment]
I can't compile the project, can't find curl/curl.h, switching to v0.6.0b9 version the problem still exists, can you compile the project properly?

**Comment by yifeikong** [comment]
You need to check out the `Makefile` and `.github/workflows` to see how the build process works.

File: curl_cffi/requests/cookies.py [hunk 3]
```diff
@@ -290,6 +290,20 @@ def delete(
...
+        requirements.
+
+        """
+        dictionary = {}
+        for cookie in self.jar:
+            if (domain is None or cookie.domain == domain) and (
+                    path is None or cookie.path == path
+            ):
+                dictionary[cookie.name] = cookie.value
+        return dictionary
```
Comment (jjsaunier) [inline curl_cffi/requests/cookies.py:305]: I don't really understand the purpose here, may be it's expected, but cookie matching is not working like that. You can check official RFC:

* https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.3 * https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.4

**Pull request closed without merging**